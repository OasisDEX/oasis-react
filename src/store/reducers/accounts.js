import { createAction, handleActions } from 'redux-actions';
import Immutable from 'immutable';

import web3 from '../../bootstrap/web3';
import { Session } from '../../utils/session';
import platformReducer from './platform';
import sessionReducer from './session';


const initialState = Immutable.fromJS({
  defaultAccount: null,
  accounts: []
});

const INIT = 'ACCOUNTS/INIT';
const CHECK_ACCOUNTS = 'NETWORK/CHECK_ACCOUNTS';

const Init = createAction(
  INIT,
  () => null,
);



const checkAccounts = createAction(
  CHECK_ACCOUNTS,
  () => window.web3p.eth.getAccounts(),
);


const checkAccountsEpic = () => async (dispatch, getState) => {
  const userAccounts = (await dispatch(checkAccounts())).value;
  const isMetamaskLocked = getState().getIn(['platform', 'metamaskLocked']);
  const hasUserAccounts = userAccounts.length;
  if(!hasUserAccounts) {
    if(!isMetamaskLocked) {
      dispatch(platformReducer.actions.metamaskLocked());
    }
    window.web3.eth.defaultAccount = undefined;
  }
  else if (!userAccounts.find( acc => acc === window.web3.eth.defaultAccount)) {
    if ((userAccounts.find( acc => acc === localStorage.getItem('address')) )) {
      window.web3.eth.defaultAccount = localStorage.getItem('address');
    } else if (userAccounts.find( acc => acc === Session.get('address'))) {
      window.web3.eth.defaultAccount = Session.get('address');
    } else if (userAccounts.length > 0) {
      window.web3.eth.defaultAccount = userAccounts[0];
    } else {
      throw new Error('No default account');
    }
  }

  if(isMetamaskLocked) {
    dispatch(platformReducer.actions.metamaskUnlocked())
  }
  localStorage.setItem('address', web3.eth.defaultAccount);
  dispatch(sessionReducer.actions.SetValue('address', window.web3.eth.defaultAccount));
  dispatch(sessionReducer.actions.SetValue('accounts', userAccounts))
};


const actions = {
  Init,
  checkAccountsEpic
};

const reducer = handleActions({}, initialState);

export default {
  actions,
  reducer,
};
