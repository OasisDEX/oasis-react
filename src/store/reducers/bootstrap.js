import { createAction, handleActions } from 'redux-actions';
import Immutable from 'immutable';

const initialState = Immutable.fromJS({});

const NETWORK_INITIALIZED = 'BOOTSTRAP/NETWORK_INITIALIZED';



const networkInitialized = createAction(
  NETWORK_INITIALIZED
);

const actions = {
  networkInitialized
};

const reducer = handleActions({}, initialState);

export default {
  actions,
  reducer,
};
