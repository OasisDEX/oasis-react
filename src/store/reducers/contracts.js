import { createAction, handleActions } from 'redux-actions';
import { fromJS } from 'immutable';
import { TOKEN_WRAPPED_GNT } from '../../constants';

const initialState = fromJS({
  implementsTokenWrappers: [
    TOKEN_WRAPPED_GNT
  ]
});

const INIT = 'CONTRACTS/INIT';

const Init = createAction(
  INIT,
  () => null,
);

const actions = {
  Init,
};

const reducer = handleActions({}, initialState);

export default {
  actions,
  reducer,
};
