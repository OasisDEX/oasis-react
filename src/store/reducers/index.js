import { combineReducers } from 'redux-immutable';
import { routerReducer } from 'react-router-redux'

import platform from './platform';
import network from './network';
import markets from './markets';
import balances from './balances';
import orders from './orders';
import history from './transferHistory';
import limits from './limits';
import offers from './offers';
import tokenEvents from './tokenEvents';
import tokens from './tokens';
import transactions from './transactions';
import weth from './weth';
import wgnt from './wgnt';
import session from './session';
import contracts from './contracts';
import accounts from './accounts';
import trades from './trades';
import widgets from './widgets';
import offerTakes from './offerTakes';
import { reducer as formReducer } from 'redux-form/immutable'
import transactionWatchers from './transactionWatchers';
import timers from './timers';
import offerMakes from './offerMakes';
import transferHistory from './transferHistory';
import transfers from './transfers';
import tokenSelectors from './tokenSelectors';
import wrapUnwrap from './wrapUnwrap';


export default combineReducers(
  {

    router: routerReducer,
    platform: platform.reducer,
    network: network.reducer,
    markets: markets.reducer,
    balances: balances.reducer,
    orders: orders.reducer,
    history: history.reducer,
    limits: limits.reducer,
    offers: offers.reducer,
    tokenEvents: tokenEvents.reducer,
    tokens: tokens.reducer,
    transactions: transactions.reducer,
    weth: weth.reducer,
    wgnt: wgnt.reducer,
    session: session.reducer,
    contracts: contracts.reducer,
    accounts: accounts.reducer,
    trades: trades.reducer,
    widgets: widgets.reducer,
    offerTakes: offerTakes.reducer,
    offerMakes: offerMakes.reducer,
    form: formReducer,
    transactionWatchers: transactionWatchers.reducer,
    timers: timers.reducer,
    transferHistory: transferHistory.reducer,
    transfers: transfers.reducer,
    tokenSelectors: tokenSelectors.reducer,
    wrapUnwrap: wrapUnwrap.reducer
  });
