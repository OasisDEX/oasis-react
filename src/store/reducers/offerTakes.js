import { createAction, handleActions } from "redux-actions";
import { fromJS } from "immutable";
import { createPromiseActions } from "../../utils/createPromiseActions";
import tokens from "../selectors/tokens";
import offerTakes from "../selectors/offerTakes";
import offersReducer from "./offers";
import { initialize } from "redux-form";
import { TX_OFFER_TAKE } from "./transactions";
import * as form from "redux-form/immutable";
import web3 from "../../bootstrap/web3";
import {
  DEFAULT_GAS_LIMIT,
  DEFAULT_GAS_PRICE,
  ETH_UNIT_ETHER
} from "../../constants";
import balances from "../selectors/balances";
import { fulfilled, pending, rejected } from "../../utils/store";
import { handleTransaction } from "../../utils/transactions/handleTransaction";
import { defer } from "../deferredThunk";
import findOffer from "../../utils/offers/findOffer";
import {
  getMarketContractInstance,
  getMarketNoProxyContractInstance
} from "../../bootstrap/contracts";
import transactions from "../selectors/transactions";
import isNumericAndGreaterThanZero from "../../utils/numbers/isNumericAndGreaterThanZero";
import { convertToTokenPrecision } from "../../utils/conversion";

export const TAKE_BUY_OFFER = "OFFER_TAKES/TAKE_BUY_OFFER";
export const TAKE_SELL_OFFER = "OFFER_TAKES/TAKE_SELL_OFFER";

const initialState = fromJS({
  activeOfferTakeType: null,
  isOfferTakeModalOpen: false,
  activeOfferTakeOfferId: null,
  minOrderLimitInWei: "100000000000000000",
  transactionGasCostEstimate: null,
  checkingIfOfferActive: null,
  isOfferActive: null,
  offerExceededGasLimit: null
});

const INIT = "OFFER_TAKES/INIT";

const Init = createAction(INIT, () => null);

const buyMaxEpic = (
  {
    quoteTokenBalance = balances.activeQuoteTokenBalance,
    offerTakeOfferData = offerTakes.activeOfferTakeOfferData
  } = {}
) => (dispatch, getState) => {
  const usersQuoteTokenBalanceBN = web3.toBigNumber(
    quoteTokenBalance(getState())
  );

  const activeOfferTakeData = offerTakeOfferData(getState());
  const volume = activeOfferTakeData.get("buyHowMuch");
  const priceBN = web3.toBigNumber(activeOfferTakeData.get("ask_price"));

  const total = usersQuoteTokenBalanceBN.gte(priceBN.mul(volume))
    ? web3.fromWei(priceBN.mul(volume), ETH_UNIT_ETHER)
    : web3.fromWei(usersQuoteTokenBalanceBN, ETH_UNIT_ETHER);

  const totalSerialized = total.toString();

  dispatch(form.change("takeOffer", "total", totalSerialized));
  dispatch(defer(totalFieldValueChangedEpic, totalSerialized));
};

const sellMaxEpic = (
  {
    activeBaseTokenBalance = balances.activeBaseTokenBalance,
    activeOfferTakeOfferData = offerTakes.activeOfferTakeOfferData
  } = {}
) => (dispatch, getState) => {
  const usersBaseTokenBalanceBN = web3.toBigNumber(
    activeBaseTokenBalance(getState())
  );
  const activeOfferTakeData = activeOfferTakeOfferData(getState());
  const volume = activeOfferTakeData.get("buyHowMuch");
  const priceBN = web3.toBigNumber(activeOfferTakeData.get("bid_price"));

  const total = usersBaseTokenBalanceBN.gte(volume)
    ? web3.fromWei(priceBN.mul(volume), ETH_UNIT_ETHER)
    : web3.fromWei(usersBaseTokenBalanceBN.mul(priceBN), ETH_UNIT_ETHER);

  const totalSerialized = web3
    .toBigNumber(total)
    .toFixed()
    .toString();
  dispatch(form.change("takeOffer", "total", totalSerialized));
  dispatch(defer(totalFieldValueChangedEpic, totalSerialized));
};

const resetCompletedOfferCheck = createAction(
  "OFFER_TAKES/RESET_COMPLETED_OFFER_CHECK"
);

const sendBuyTransaction = createAction(
  "OFFER_TAKES/MARKET_BUY",
  (
    offerId,
    amount,
    gasPrice = DEFAULT_GAS_PRICE
  ) =>
    getMarketContractInstance().buy(offerId, amount, {
      gasPrice
    })
);

const takeOffer = createPromiseActions("OFFER_TAKES/TAKE_OFFER");

const takeOfferEpic = (withCallbacks = {}) => async (dispatch, getState) => {
  const amountInWei = convertToTokenPrecision(
    web3.toWei(offerTakes.getBuyAmount(getState()), ETH_UNIT_ETHER),
    offerTakes.activeOfferTakeOfferData(getState()).get("sellWhichToken")
  );

  const activeOfferTakeOfferId = offerTakes.activeOfferTakeOfferId(getState());

  dispatch(takeOffer.pending());

  return handleTransaction({
    dispatch,
    transactionDispatcher: () =>
      dispatch(sendBuyTransaction(activeOfferTakeOfferId, amountInWei)),
    transactionType: TX_OFFER_TAKE,
    withCallbacks
  });
};

const setActiveOfferTakeType = createAction(
  "OFFER_TAKES/SET_ACTIVE_OFFER_TAKE_TYPE",
  takeType => takeType
);
const resetActiveOfferTakeType = createAction(
  "OFFER_TAKES/RESET_ACTIVE_OFFER_TAKE_TYPE"
);

const setOfferTakeModalOpen = createAction("OFFER_TAKES/SET_MODAL_OPEN");
const setOfferTakeModalOpenEpic = ({ offerId, offerTakeType }) => dispatch => {
  dispatch(setActiveOfferTakeOfferId(offerId));
  dispatch(setActiveOfferTakeType(offerTakeType));
  dispatch(defer(initializeOfferTakeFormEpic));
  dispatch(defer(getTransactionGasCostEstimateEpic));
  dispatch(setOfferTakeModalOpen());
};

const setOfferTakeModalClosed = createAction("OFFER_TAKES/SET_MODAL_CLOSED");
const setOfferTakeModalClosedEpic = () => dispatch => {
  dispatch(setOfferTakeModalClosed());
  dispatch(resetActiveOfferTakeType());
  dispatch(resetActiveOfferTakeOfferId());
  dispatch(resetActiveOfferTakeGasCostEstimate());
  dispatch(resetCompletedOfferCheck());
};

const initializeOfferTakeForm = createAction(
  "OFFER_TAKES/INITIALIZE_OFFER_TAKE_FORM"
);
const initializeOfferTakeFormEpic = () => (dispatch, getState) => {
  const activeOfferTake = offerTakes.activeOfferTake(getState());
  const activeTradingPairPrecision = tokens.precision(getState());
  let price, volume, total;
  switch (offerTakes.activeOfferTakeType(getState())) {
    case TAKE_SELL_OFFER:
      [price, volume, total] = [
        activeOfferTake.getIn(["offerData", "ask_price"]),
        activeOfferTake.getIn(["offerData", "sellHowMuch"]),
        activeOfferTake.getIn(["offerData", "buyHowMuch"])
      ];
      break;
    case TAKE_BUY_OFFER:
      [price, volume, total] = [
        activeOfferTake.getIn(["offerData", "bid_price"]),
        activeOfferTake.getIn(["offerData", "buyHowMuch"]),
        activeOfferTake.getIn(["offerData", "sellHowMuch"])
      ];
      break;
    default:
      throw new Error("No offer take type provided!");
  }
  dispatch(
    initialize("takeOffer", {
      price,
      volume: web3
        .fromWei(volume, ETH_UNIT_ETHER, {
          precision: activeTradingPairPrecision
        })
        .toString(),
      total: web3
        .fromWei(total, ETH_UNIT_ETHER, {
          precision: activeTradingPairPrecision
        })
        .toString()
    })
  );
  dispatch(initializeOfferTakeForm());
};

const setActiveOfferTakeOfferId = createAction(
  "OFFER_TAKES/SET_ACTIVE_OFFER_TAKE_OFFER_ID"
);
const resetActiveOfferTakeOfferId = createAction(
  "OFFER_TAKES/RESET_ACTIVE_OFFER_TAKE_OFFER_ID"
);

const checkIfOfferTakeSubjectIsActive = createPromiseActions(
  "OFFER_TAKES/CHECK_IF_OFFER_TAKE_SUBJECT_IS_ACTIVE"
);
const checkIfOfferTakeSubjectStillActiveEpic = offerId => async (
  dispatch,
  getState
) => {
  const activeOfferTakeOfferId =
    offerId || offerTakes.activeOfferTakeOfferId(getState());
  dispatch(checkIfOfferTakeSubjectIsActive.pending());
  const isActive = (await dispatch(
    offersReducer.actions.checkOfferIsActive(activeOfferTakeOfferId)
  )).value;
  dispatch(checkIfOfferTakeSubjectIsActive.fulfilled(isActive));
  if (false === isActive) {
    dispatch(
      offersReducer.actions.markOfferAsInactive(
        findOffer(activeOfferTakeOfferId, getState())
      )
    );
  }
  return isActive;
};

const volumeFieldValueChangedEpic = (
  value,
  { formValueSelector = form.formValueSelector } = {}
) => (dispatch, getState) => {
  const { price } = formValueSelector("takeOffer")(
    getState(),
    "volume",
    "total",
    "price"
  );

  if (isNaN(value)) {
    dispatch(form.change("takeOffer", "total", "0"));
  } else {
    dispatch(
      form.change(
        "takeOffer",
        "total",
        web3
          .toBigNumber(value)
          .mul(price)
          .toFixed()
          .toString()
      )
    );
  }

  dispatch(defer(getTransactionGasCostEstimateEpic));
};

const totalFieldValueChangedEpic = (
  value,
  { formValueSelector = form.formValueSelector } = {}
) => (dispatch, getState) => {
  const { price } = formValueSelector("takeOffer")(
    getState(),
    "volume",
    "total",
    "price"
  );

  dispatch(
    form.change(
      "takeOffer",
      "volume",
      web3
        .toBigNumber(value)
        .div(price)
        .toFixed()
        .toString()
    )
  );

  if (isNaN(value)) {
    dispatch(form.change("takeOffer", "volume", "0"));
  } else {
    dispatch(
      form.change(
        "takeOffer",
        "volume",
        web3
          .toBigNumber(value)
          .div(price)
          .toFixed()
          .toString()
      )
    );
  }

  dispatch(defer(getTransactionGasCostEstimateEpic));
};

const resetActiveOfferTakeGasCostEstimate = createAction(
  "OFFER_TAKES/RESET_TRANSACTION_GAS_ESTIMATE"
);

const getTransactionGasCostEstimate = createAction(
  "OFFER_TAKES/GET_TRANSACTION_GAS_ESTIMATE",
  ({ offerId, amount, offerOwner, activeOfferData }) =>
    new Promise((resolve, reject) => {
      console.log(
        "trying to estimate: ",
        offerId,
        amount,
        offerOwner,
        DEFAULT_GAS_LIMIT
      );

      getMarketNoProxyContractInstance().buy.estimateGas(
        offerId,
        amount,
        { to: offerOwner },
        (e, estimation) => {
          if (e) {
            console.log("gas estimation failed!");
            reject({
              offerId,
              amount,
              to: offerOwner,
              activeOfferData
            });
          } else {
            console.log(
              "gas estimation:",
              offerId,
              amount,
              offerOwner,
              estimation
            );
            resolve(estimation);
          }
        }
      );
    })
);

const getTransactionGasCostEstimateEpic = (
  {
    canFulfillOffer = offerTakes.canFulfillOffer,
    activeOfferTakeOfferId = offerTakes.activeOfferTakeOfferId,
    takeFormValuesSelector = offerTakes.takeFormValuesSelector,
    activeOfferTakeOfferOwner = offerTakes.activeOfferTakeOfferOwner,
    activeOfferTakeOfferData = offerTakes.activeOfferTakeOfferData,
    getBuyAmount = offerTakes.getBuyAmount
  } = {}
) => async (dispatch, getState) => {
  const offerId = activeOfferTakeOfferId(getState());
  const { volume, total } = takeFormValuesSelector(
    getState(),
    "volume",
    "total"
  );

  if (
    !isNumericAndGreaterThanZero(volume) ||
    !isNumericAndGreaterThanZero(total) ||
    !canFulfillOffer(getState())
  ) {
    return null;
  }

  // console.log("===========================================================================================");
  // console.log("transactionGasCostEstimate",
  //   getBuyAmount(getState()),
  //   offerTakes.activeOfferTakeType(getState()),
  //   activeOfferTakeOfferData(getState()).get("buyWhichToken"),
  //   activeOfferTakeOfferData(getState()).get("sellWhichToken"),
  //   offerTakes.activeOfferTakeType(getState()) === TAKE_BUY_OFFER
  //   ? activeOfferTakeOfferData(getState()).get("buyWhichToken")
  //   : activeOfferTakeOfferData(getState()).get("sellWhichToken"));
  // console.log("===========================================================================================");
  
  const transactionGasCostEstimate = (await dispatch(
    defer(getTransactionGasCostEstimate, {
      offerId,
      amount: convertToTokenPrecision(
        web3.toWei(getBuyAmount(getState()), ETH_UNIT_ETHER),
        activeOfferTakeOfferData(getState()).get("sellWhichToken")
      ),
      offerOwner: activeOfferTakeOfferOwner(getState()),
      activeOfferData: activeOfferTakeOfferData(getState())
    })
  )).value;

  const gasLimitInWeiBN = web3.toBigNumber(
    transactions.defaultGasLimit(getState())
  );
  if (gasLimitInWeiBN.lt(transactionGasCostEstimate)) {
    dispatch(setGasExceedsTheLimitEnabled());
  } else {
    dispatch(setGasExceedsTheLimitDisabled());
  }
  return transactionGasCostEstimate;
};

const setGasExceedsTheLimitEnabled = createAction(
  "OFFER_TAKES/GAS_EXCEEDS_THE_LIMIT_ENABLED"
);
const setGasExceedsTheLimitDisabled = createAction(
  "OFFER_TAKES/GAS_EXCEEDS_THE_LIMIT_DISABLED"
);

const actions = {
  Init,
  takeOffer,
  takeOfferEpic,
  setOfferTakeModalClosedEpic,
  setOfferTakeModalOpenEpic,
  setActiveOfferTakeType,
  setActiveOfferTakeOfferId,
  resetActiveOfferTakeOfferId,
  checkIfOfferTakeSubjectStillActiveEpic,
  volumeFieldValueChangedEpic,
  totalFieldValueChangedEpic,
  buyMaxEpic,
  sellMaxEpic,
  getTransactionGasCostEstimateEpic,
  resetCompletedOfferCheck
};

const reducer = handleActions(
  {
    [setOfferTakeModalOpen]: state => state.set("isOfferTakeModalOpen", true),
    [setOfferTakeModalClosed]: state =>
      state.set("isOfferTakeModalOpen", false),
    [setActiveOfferTakeType]: (state, { payload }) =>
      state.set("activeOfferTakeType", payload),
    [resetActiveOfferTakeType]: state => state.set("activeOfferTakeType", null),
    [setActiveOfferTakeOfferId]: (state, { payload }) =>
      state.set("activeOfferTakeOfferId", payload),
    [resetActiveOfferTakeOfferId]: state =>
      state.set("activeOfferTakeOfferId", null),
    [checkIfOfferTakeSubjectIsActive.pending]: state =>
      state.set("checkingIfOfferActive", true),
    [checkIfOfferTakeSubjectIsActive.fulfilled]: (state, { payload }) =>
      state.set("isOfferActive", payload).set("checkingIfOfferActive", false),
    [pending(getTransactionGasCostEstimate)]: state =>
      state
        .set("transactionGasCostEstimateError", false)
        .set("transactionGasCostEstimatePending", true),
    [fulfilled(getTransactionGasCostEstimate)]: (state, { payload }) =>
      state
        .set("transactionGasCostEstimateError", false)
        .set("transactionGasCostEstimate", payload.toString())
        .set("transactionGasCostEstimatePending", false),
    [rejected(getTransactionGasCostEstimate)]: state =>
      state
        .set("transactionGasCostEstimateError", true)
        .set("transactionGasCostEstimatePending", false),
    [resetActiveOfferTakeGasCostEstimate]: state =>
      state
        .set("transactionGasCostEstimate", null)
        .set("transactionGasCostEstimateError", null)
        .set("transactionGasCostEstimatePending", null),

    [resetCompletedOfferCheck]: state =>
      state.set("checkingIfOfferActive", null).set("isOfferActive", null),
    [setGasExceedsTheLimitEnabled]: state =>
      state.set("exceededGasLimit", true),
    [setGasExceedsTheLimitDisabled]: state =>
      state.set("exceededGasLimit", false)
  },
  initialState
);

export default {
  actions,
  reducer
};
