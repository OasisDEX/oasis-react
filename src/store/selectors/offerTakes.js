import { createSelector } from "reselect";
import { formValueSelector } from "redux-form/immutable";
import balances from "./balances";
import transactions from "./transactions";
import markets from "./markets";
import web3 from "../../bootstrap/web3";
import { TAKE_BUY_OFFER, TAKE_SELL_OFFER } from "../reducers/offerTakes";
import { ETH_UNIT_ETHER } from "../../constants";
import { fromJS } from "immutable";
import getOfferTakeBuyAndSellTokens from "../../utils/tokens/getOfferTakeBuyAndSellTokens";
import tokens from "./tokens";
import tokenToBeAllowedForOffer from "../../utils/offers/tokenToBeAllowedForOffer";
import offers from "./offers";

const offerTakesState = s => s.get("offerTakes");

const activeOfferTakeType = createSelector(offerTakesState, s =>
  s.get("activeOfferTakeType")
);

const isOfferTakeModalOpen = createSelector(offerTakesState, s =>
  s.get("isOfferTakeModalOpen")
);
const activeOfferTakeOfferId = createSelector(offerTakesState, s =>
  s.get("activeOfferTakeOfferId")
);

const activeOfferTake = createSelector(
  offers.activeTradingPairBuyOffers,
  offers.activeTradingPairSellOffers,
  activeOfferTakeType,
  activeOfferTakeOfferId,
  tokens.activeTradingPair,
  (activeTradingPairBuyOffers, activeTradingPairSellOffers, offerTakeType, offerId, activeTradingPair) => {
    const { baseToken, quoteToken } = activeTradingPair;
    let offer = null;
    switch (offerTakeType) {
      case TAKE_BUY_OFFER:
        {
          offer = activeTradingPairBuyOffers
            .find(offer => fromJS(offer).get("id") === offerId);
        }
        break;
      case TAKE_SELL_OFFER:
        {
          offer = activeTradingPairSellOffers
            .find(offer => fromJS(offer).get("id") === offerId);
        }
        break;
    }

    if (offer) {
      const { sellToken, buyToken } = getOfferTakeBuyAndSellTokens(
        activeTradingPair,
        offerTakeType
      );
      return fromJS({
        offerData: fromJS(offer),
        sellToken: sellToken,
        buyToken: buyToken,
        baseToken: baseToken,
        quoteToken: quoteToken
      });
    } else {
      return fromJS({});
    }
  }
);

const activeOfferTakeOfferData = createSelector(activeOfferTake, s =>
  s.get("offerData")
);

const activeOfferTakeOfferOwner = createSelector(activeOfferTake, s =>
  s.getIn(["offerData", "owner"])
);

const activeOfferTakeBuyToken = createSelector(activeOfferTake, s =>
  s.get("buyToken")
);

const activeOfferTakeSellToken = createSelector(activeOfferTake, s =>
  s.get("sellToken")
);

const takeFormValuesSelector = formValueSelector("takeOffer");

const isOfferBelowLimit = createSelector(offerTakesState, s =>
  s.get("isOfferBelowLimit")
);

const hasSufficientTokenAmount = createSelector(
  balances.tokenBalances,
  activeOfferTakeSellToken,
  activeOfferTakeType,
  rootState => takeFormValuesSelector(rootState, "volume", "total", "price"),
  (tokenBalances, sellToken, activeOfferTakeType, { total, volume }) => {
    if (!total) {
      return true;
    } else {
      const totalInWei = web3.toWei(total, ETH_UNIT_ETHER);
      const volumeInWei = web3.toWei(volume, ETH_UNIT_ETHER);
      const usersSellTokenBalanceBN = web3.toBigNumber(
        tokenBalances.get(sellToken)
      );
      switch (activeOfferTakeType) {
        case TAKE_BUY_OFFER:
          return usersSellTokenBalanceBN.gte(volumeInWei);
        case TAKE_SELL_OFFER:
          return usersSellTokenBalanceBN.gte(totalInWei);
      }
    }
  }
);

const currentFormVolume = createSelector(
  rootState => takeFormValuesSelector(rootState, "volume"),
  volume => volume
);

const isVolumeEmptyOrZero = createSelector(
  rootState => takeFormValuesSelector(rootState, "volume"),
  activeFormVolume => {
    return !activeFormVolume || web3.toBigNumber(activeFormVolume).eq(0);
  }
);

const isVolumeGreaterThanOfferMax = createSelector(
  currentFormVolume,
  activeOfferTakeOfferData,
  activeOfferTakeType,
  (volume, offerData, offerTakeType) => {
    if (!offerData) {
      return false;
    }
    const buyHowMuch = web3.fromWei(offerData.get("buyHowMuch"));
    const sellHowMuch = web3.fromWei(offerData.get("sellHowMuch"));
    const volumeBN = web3.toBigNumber(volume);
    switch (offerTakeType) {
      case TAKE_BUY_OFFER:
        return volumeBN.gt(buyHowMuch) ? buyHowMuch : false;
      case TAKE_SELL_OFFER:
        return volumeBN.gt(sellHowMuch) ? sellHowMuch : false;
    }
  }
);

const hasExceededGasLimit = createSelector(offerTakesState, s =>
  Boolean(s.get("exceededGasLimit"))
);

const tokenToBeAllowed = createSelector(
  activeOfferTakeType,
  activeOfferTakeSellToken,
  activeOfferTakeBuyToken,
  (offerType, sellToken, buyToken) => {
    return tokenToBeAllowedForOffer({ offerType, sellToken, buyToken });
  }
);

const isOfferActive = createSelector(offerTakesState, s =>
  s.get("isOfferActive")
);


const canFulfillOffer = createSelector(
  hasSufficientTokenAmount,
  rootState => transactions.canSendTransaction(rootState),
  markets.isBuyEnabled,
  rootState => {
    return balances.tokenAllowanceStatusForActiveMarket(rootState, {
      tokenName: tokenToBeAllowed(rootState)
    });
  },
  isVolumeEmptyOrZero,
  isVolumeGreaterThanOfferMax,
  isOfferActive,
  (
    hasSufficientTokenAmount,
    canSendTransaction,
    isBuyEnabled,
    istTokenEnabled,
    isVolumeZero,
    isVolumeGreaterThanOfferMax,
  ) => {
    if (
      !isOfferActive ||
      isVolumeZero ||
      isVolumeGreaterThanOfferMax ||
      // !canSendTransaction ||
      !isBuyEnabled ||
      [false, null].includes(istTokenEnabled)
    ) {
      return false;
    } else {
      return hasSufficientTokenAmount;
    }
  }
);

const transactionGasCostEstimate = createSelector(offerTakesState, s =>
  s.get("transactionGasCostEstimate")
);

const checkingIfOfferIsActive = createSelector(offerTakesState, s =>
  s.get("checkingIfOfferActive")
);

const gasEstimatePending = createSelector(offerTakesState, s =>
  s.get("transactionGasCostEstimatePending")
);

const getActiveOfferTakeAllowanceStatus = createSelector(
  rootState => {
    return balances.tokenAllowanceStatusForActiveMarket(rootState, {
      tokenName: tokenToBeAllowed(rootState)
    });
  },
  status => Boolean(status)
);

const isActiveOfferTakeBestOffer = createSelector(
  activeOfferTakeType,
  activeOfferTakeOfferId,
  offers.activeTradingPairBestBuyOfferId,
  offers.activeTradingPairBestSellOfferId,
  (takeType, offerId, bestBuyOfferId, bestSellOfferId) => {
    switch (takeType) {
      case TAKE_BUY_OFFER:
        return offerId.toString() === bestBuyOfferId;
      case TAKE_SELL_OFFER:
        return offerId.toString() === bestSellOfferId;
    }
  }
);

const getBuyAmount = createSelector(
  activeOfferTakeType,
  activeOfferTakeBuyToken,
  activeOfferTakeSellToken,
  rootState => rootState,
  (activeOfferType, buyToken, sellToken, rs) =>
    activeOfferType === TAKE_BUY_OFFER
      ? takeFormValuesSelector(rs, "total")
      : takeFormValuesSelector(rs, "volume")
);

export default {
  state: offerTakesState,
  activeOfferTake,
  activeOfferTakeType,
  activeOfferTakeOfferData,
  isOfferTakeModalOpen,
  activeOfferTakeOfferId,
  activeOfferTakeBuyToken,
  activeOfferTakeSellToken,
  activeOfferTakeOfferOwner,
  canFulfillOffer,
  isOfferBelowLimit,
  takeFormValuesSelector,
  hasSufficientTokenAmount,
  transactionGasCostEstimate,
  isVolumeGreaterThanOfferMax,
  isVolumeEmptyOrZero,
  isOfferActive,
  checkingIfOfferIsActive,
  gasEstimatePending,
  getActiveOfferTakeAllowanceStatus,
  isActiveOfferTakeBestOffer,
  tokenToBeAllowed,
  getBuyAmount,
  hasExceededGasLimit
};
