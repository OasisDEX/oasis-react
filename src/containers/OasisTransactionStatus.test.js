/* global shallow describe it expect */
/* eslint-disable import/first */
import React from 'react';
import { fromJS } from 'immutable';

import {
  OasisTransactionStatusWrapper,
  mapStateToProps,
  mapDispatchToProps
} from './OasisTransactionStatus';
import { shallow } from 'enzyme';
import { TAKE_BUY_OFFER } from '../store/reducers/offerTakes';

describe('(Container) OasisTransactionStatus', () => {
  const state = fromJS(global.storeMock);
  const initialProps = mapStateToProps(state, {});
  const initialActions = mapDispatchToProps(x => x);
  const props = {
    ...initialActions,
    ...initialProps,
    txTimestamp: 1,
    transaction: fromJS({txType: TAKE_BUY_OFFER})
  };

  it('will receive right props', () => {
    expect(initialProps).toMatchSnapshot();
  });


  it('will receive right actions', () => {
    expect(initialActions).toMatchSnapshot();
  });

  it('should render', () => {
    const wrapper = shallow(
      <OasisTransactionStatusWrapper {...props}/>
    );
    expect(wrapper).toMatchSnapshot();
  });

});
