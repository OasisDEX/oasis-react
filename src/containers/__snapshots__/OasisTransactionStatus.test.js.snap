// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Container) OasisTransactionStatus should render 1`] = `
<InfoBox
  alignItems="baseline"
  className=""
  color="default"
  fullWidth={true}
  justifyContent="space-between"
  size="sm"
>
  <div
    className=""
    style={
      Object {
        "alignContent": "space-between",
        "display": "flex",
      }
    }
  >
    <div />
    <TransactionStatus
      noBorder={true}
      transaction={
        Immutable.Map {
          txType: "OFFER_TAKES/TAKE_BUY_OFFER",
        }
      }
    />
  </div>
</InfoBox>
`;

exports[`(Container) OasisTransactionStatus will receive right actions 1`] = `
Object {
  "actions": Object {},
}
`;

exports[`(Container) OasisTransactionStatus will receive right props 1`] = `
Object {
  "transaction": Immutable.Map {
    txStatus: undefined,
},
}
`;
